src/client/errors.ts

```diff
@@ -26,11 +26,17 @@ class ErrorClientInvalidNode extends ErrorClient {
   exitCode: number = 70;
 }
 
+class ErrorClientInvalidProvider extends ErrorClient {
```

Comment:
What is this exception for? In our `IdentitiesManager` we return `undefined` when we cannot provide a given provider. Unless we have a call that expects a provider to exist, we can propagate this value-error via GRPC in a empty message sort of value. That is if this is used to fetch a provider, you would return undefined/null instead of throwing an error. The caller must then check if it actually got the data or not.

In grpc I'm pretty we have optional message properties.

Reply:
Same goes for all other kinds of GRPC calls that fetch things. You don't use an exception to indicate that it doesn't exist. You use an undefined/null value instead.

Now when do you use such an exception, you do use it when there will be a call that takes a provider id or whatever id and you intend to do something with that id. Then in that case, an exception can be thrown. But make sure that such exceptions are isolated to those calls, and not calls that simply fetch a resource.

Reply:
It's for when the user passes in an invalid provider as the argument for a CLI call, so it's more of a client-side error. Since the command can't continue if the provided input is incorrect we need to throw an error.

Reply:
Yea but you must differentiate between these commands:

1. Commands that get a resource based on the ID. The response is not an error, simply an undefined/null.
2. Commands that do something to a resource based on the ID. The response is an error because the ID doesn't exist.

Special case:

3. Commands that get a resource B that depends on resource A that is indexed by the ID. In this case it could be either. But I usually default on returning undefined because the transitive lookup failed.

Reply:
I should clarify the above. The above depends on API context.

In case 1, you usually have a `404` to represent a client error when a resource is missing in HTTP. But in RPC and functional calls, this should be translated to a `Maybe` type which results in either a value or a sentinel value representing emptiness like `null` or `undefined.

If you're in a CLI environment, you should probably throw an error here with an exit code. You don't want to just return nothing with exit code 0. EXCEPT when the command is meant to return a series of data. In such a case, a succesful exit code with 0 results is sufficient.

So for example:

```sh
# imagine we are in the terminal
get-one-thing <id>
# it should be exit code > 0 because the id doesn't exsit

get-many-things/search-for-things <id>
# it should be exit code 0 with no results because it is successfully getting you 0 results
```

This also applies to network APIs.

Reply:
Ok I have written up a guideline here: https://github.com/MatrixAI/js-polykey/wiki/API-Design. Let me know if that helps.

Reply:
The client errors that were previously being used (`ErrorClientInvalidNode` and `ErrorClientInvalidProvider`) have been removed from everywhere except for two places:
- `identitiesClaim` - we need to throw an error if the provider is invalid before we generate the claim on our sigchain, because once we do that we can't remove it, and if the provider is invalid we won't be able to publish the claim on the identity
- `identitiesInfoGetConnected` - this rpc doesn't have a corresponding bin command so I wasn't sure what its purpose was. Should this even exist?

Reply:
Seems right for `identitiesClaim`, that's one of the cases where you are doing something to a resource with the expectation that the ID is valid.

`identitiesInfoGetConnected` - Does this get a transitive resource? This is the ambiguous one. Sometimes we would say that the ID isn't valid, so you have no connections. But I think in this case, we would want to differentiate the fact that the identity doesn't actually exist, vs the identity exist, but empty list of connections. Therefore it should throw an error.

Reply:
The thing with `identitiesInfoGetConnected` is that it isn't used anywhere in the CLI. There's a test for it in `rpcIdentities.test.ts` but other than that it's not used anywhere in the codebase.